antsibull/build_changelog.py

```diff
@@ -0,0 +1,504 @@
+
+
+def insert_after_heading(lines: t.List[str], content: str) -> None:
+    has_heading = False
+    for index, line in enumerate(lines):
+        if line.startswith('***') and line == '*' * len(line):
+            has_heading = True
+        elif has_heading:
+            if line:
+                has_heading = False
+            else:
+                # First empty line after top-level heading: insert TOC
+                lines.insert(index, content)
+                return
+
+
+#
+# Release Notes
+#
+
+
+class ReleaseNotes:
+    changelog_filename: str
+    changelog_bytes: bytes
+
+    porting_guide_filename: str
+    porting_guide_bytes: bytes
+
+    def __init__(self, changelog_filename: str, changelog_bytes: bytes,
+                 porting_guide_filename: str, porting_guide_bytes: bytes):
+        self.changelog_filename = changelog_filename
+        self.changelog_bytes = changelog_bytes
+
+        self.porting_guide_filename = porting_guide_filename
+        self.porting_guide_bytes = porting_guide_bytes
+
+    @staticmethod
+    def _get_changelog_bytes(changelog: Changelog) -> bytes:
+        builder = RstBuilder()
+        builder.set_title(
+            f"Ansible {changelog.acd_version.major}.{changelog.acd_version.minor} Release Notes")
+        builder.add_raw_rst('.. contents::\n  :local:\n  :depth: 2\n')
+
+        for changelog_entry in changelog.entries:
+            append_changelog(builder, changelog_entry)
+
+        return builder.generate().encode('utf-8')
+
+    @staticmethod
+    def _get_porting_guide_bytes(changelog: Changelog) -> bytes:
+        builder = RstBuilder()
+        base_porting_guide = changelog.base_collector.porting_guide
+        if base_porting_guide:
+            lines = base_porting_guide.decode('utf-8').splitlines()
+            lines.append('')
+            # insert_after_heading(lines, '\n.. contents::\n  :local:\n  :depth: 2')
+            for line in lines:
+                builder.add_raw_rst(line)
+        else:
+            builder.add_raw_rst(
+                f".. _porting_{changelog.acd_version.major}."
+                f"{changelog.acd_version.minor}_guide:\n")
+            builder.set_title(
+                f"Ansible {changelog.acd_version.major}.{changelog.acd_version.minor}"
+                f" Porting Guide")
+            builder.add_raw_rst('.. contents::\n  :local:\n  :depth: 2\n')
+
+        for porting_guide_entry in changelog.entries:
+            append_porting_guide(builder, porting_guide_entry)
+
+        return builder.generate().encode('utf-8')
+
+    @staticmethod
+    def build(changelog: Changelog) -> 'ReleaseNotes':
+        return ReleaseNotes(
+            f"CHANGELOG-v{changelog.acd_version.major}.{changelog.acd_version.minor}.rst",
+            ReleaseNotes._get_changelog_bytes(changelog),
+            f"porting_guide_{changelog.acd_version.major}.{changelog.acd_version.minor}.rst",
+            ReleaseNotes._get_porting_guide_bytes(changelog),
```

Comment:
Maybe this should be a `@classmethod` and then the return can be:

``` python
return cls(
            f"CHANGELOG-v{changelog.acd_version.major}.{changelog.acd_version.minor}.rst",
            cls._get_changelog_bytes(changelog),
            f"porting_guide_{changelog.acd_version.major}.{changelog.acd_version.minor}.rst",
            cls._get_porting_guide_bytes(changelog),
)
```

Reply:
(It seems like an alternative "constructor" (python `__init__()` is actually an initializer so it's technically an alternate initializer) and constructors are classmethods)

Reply:
Changed in 6ff5a697b319aedf98b6fcad8b629df6df77f560.